### PR TITLE
Fix #409: Allow ignoring checked exceptions while waiting

### LINKED ISSUE #409: Allow ignoring checked exceptions while waiting
When using `FluentWait`, it is possible to ignore specified `RuntimeExceptions`. Checked exceptions can not be used since there is no *predicate* accepted by `until` methods that allows to throw checked exception from `apply` method. I see several ways to implement that:

1. Introduce new predicate type that permits checked exceptions.
2. Introduce Wrapping `RuntimeExcepiton` predicate implementations will use to propagate checked exceptions. `FluentWait#propagateIfNotIngored` will then unwrap before performing the check.
3. Make `FluentWait#propagateIfNotIngored` protected so subclasses can do 2.

To me, 1. seems preferable.

### PR DESCRIPTION
**Pull request by olivergondza**
Proposing fix for #409. (I have submitted my CLA today.)

Note that I have no experience with used build tools, though what I have done seems working.

---
This change is [Reviewable](https://reviewable.io/reviews/seleniumhq/selenium/414)

### CODE REVIEW

File: java/client/src/org/openqa/selenium/support/ui/FluentWait.java [hunk 1]
```diff
@@ -172,11 +172,32 @@ public FluentWait(T input, Clock clock, Sleeper sleeper) {
    * @throws TimeoutException If the timeout expires.
    */
   public void until(final Predicate<T> isTrue) {
-    until(new Function<T, Boolean>() {
+    until(new CheckedFunction<T, Boolean>() {
       public Boolean apply(T input) {
```
Comment (jleyba) [inline java/client/src/org/openqa/selenium/support/ui/FluentWait.java]: Add missing `@Override`

File: java/client/src/org/openqa/selenium/support/ui/FluentWait.java [hunk 2]
```diff
@@ -172,11 +172,32 @@ public FluentWait(T input, Clock clock, Sleeper sleeper) {
...
+  /**
+   * Repeatedly applies this instance's input value to the given predicate until the timeout expires
+   * or the predicate evaluates to true.
+   *
+   * @param isTrue The predicate to wait on.
+   * @throws TimeoutException If the timeout expires.
+   */
+  public <V> V until(final Function<? super T, V> isTrue) {
+    return until(new CheckedFunction<T, V>() {
+      public V apply(T input) {
```
Comment (jleyba) [inline java/client/src/org/openqa/selenium/support/ui/FluentWait.java]: Add missing `@Override`

File: java/client/test/org/openqa/selenium/support/ui/FluentWaitTest.java [hunk 3]
```diff
@@ -258,6 +259,32 @@ protected RuntimeException timeoutException(String message, Throwable lastExcept
...
+    when(mockClock.laterBy(0L)).thenReturn(2L);
+    when(mockClock.isNowBefore(2L)).thenReturn(false);
+
+    FluentWait<WebDriver> wait = new FluentWait<WebDriver>(mockDriver, mockClock, mockSleeper)
+        .withTimeout(0, TimeUnit.MILLISECONDS)
+        .pollingEvery(2, TimeUnit.SECONDS)
+        .ignoring(IOException.class);
+
+    CheckedFunction<WebDriver, Boolean> pred = new CheckedFunction<WebDriver, Boolean>() {
+        public Boolean apply(WebDriver input) throws Exception {
```
Comment (jleyba) [inline java/client/test/org/openqa/selenium/support/ui/FluentWaitTest.java]: 2 space indent here and below.

Also, add missing `@Override`

File: java/client/src/org/openqa/selenium/support/ui/CheckedFunction.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+/*
```
Comment (jleyba) [inline java/client/src/org/openqa/selenium/support/ui/CheckedFunction.java]: This is the wrong copyright notice. See CONTRIBUTING.md (or just copy from another java file, they've all been updated)

File: java/client/src/org/openqa/selenium/support/ui/CheckedFunction.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+ */
+
+package org.openqa.selenium.support.ui;
+
+/**
+ * Single argument predicate allowed to throw checked exceptions.
```
Comment (jleyba) [inline java/client/src/org/openqa/selenium/support/ui/CheckedFunction.java]: This is only a predicate in the context of `FluentWait`. Change to

```
A function that transforms an input value into another.
```

**Comment by olivergondza** [comment]
Rebased on master.

File: java/client/src/org/openqa/selenium/support/ui/CheckedFunction.java (hunk 4 above)
Comment (olivergondza) [inline java/client/src/org/openqa/selenium/support/ui/CheckedFunction.java]: Fixed by the rebase, github just does not realized as this line has not changed.

**Comment by olivergondza** [comment]
Is there anything else I can do to have this merged?

**Comment by olivergondza** [comment]
Is there a real interest to integrate this functionality?

**Pull request closed without merging**